module/zfs/dmu_send.c

```diff
@@ -1942,6 +1947,13 @@ setup_featureflags(struct dmu_send_params *dspp, objset_t *os,
 		*featureflags |= DMU_BACKUP_FEATURE_LZ4;
 	}
 
+	if ((*featureflags &
```

Comment:
How do we want to deal with backwards compatibility here?

If you are using 'zfs send -e' (embedded block pointers), this will end up making a send stream that cannot be received by versions of ZFS that do not yet support ZSTD. Replication streams have always been backwards compatible, and new features activated by passing specific flags.

This one is kind of a unique case, as it somewhat changes the behaviour of an existing flag (-e), to make it also require a newer feature. I am not sure there is much of a way around it, because of the way embedded block pointers work. It may just be that we need to document this somewhere clearly, maybe even adding a warning on STDERR when you use -e and have zstd activated.

Reply:
At least an error should be thrown and relatively easy to add.
And when it comes to compatibility... I think this might require us to go quite out-of-scope to fix....

Reply:
So, when we set the DMU_BACKUP_FEATURE_ZSTD flag, it will cause the older received (that does not support ZSTD) to report that the stream contains an unknown feature flag. The fact that this happens when you use -e, even if none of the datasets being transferred use ZSTD, is likely to cause confusion.

However, since the flags must be set at the start of the stream, I am not sure what the best way around it is.

Reply:
What we could do (but not retroactively), is a way of flag override "Accept unknown feature flag mismatch"... Something like an `-f`

Reply:
That won't work properly when there actually are zstd compressed (embedded) blocks.

Looking at this more closely, it uses feature_is_active, not just feature is supported, so it won't set the zstd flag unless you have a zstd compressed dataset, so we likely do not need to make any changes.

We might want to document that you need to NOT set -e if you want to `zfs send` to a non-zstd pool.

Reply:
@c0d3z3r0 You on this?

Reply:
@allanjude @Ornias1993 what I understood now is that we do not want to add a specific error message here but instead add this to the `zstd send` man page. correct?

Reply:
@c0d3z3r0 
You can do both:
- Add it to manpage
- Add an error "if you try and send to a non-zstd pool, you often do not need to set -e"

Reply:
hmm, we need to add the error on the receiver side, don't we?

Reply:
In the problem case, the receiver will be older, and won't have any new code you add. So that won't help us here.

The receiver will say 'this replication stream uses an unknown feature flag', so the receiver will get the correct error.

We may want to add a warning when we set this stream-feature flag, because of the 'policy of least astonishment'. Specifically, if the sender did not set the -c flag, in all previous versions of ZFS this would mean the receiver did not need to support the compression algorithm used, since the blocks would be sent uncompressed. The -e flag only required the receiver to support 'embedded block pointers' (which implied support for LZ4, but that was not really a consideration at the time). Now using -e ALSO implies support for zstd. Hense why we want to warn the user that -e + zstd = possibly incompatible stream.

I am not sure if we really want to add a nag warning, but it feels like just burying this information in a man page may not be enough. It might be a godo topic to bring up in next weeks ZFS Leadership call, to get a wider consensus.

Reply:
I think a nag warning in this case is totally warranted.
Considering this isn't a "niche" problem, but something we actually expect people to run into and start filing issues about.

Reply:
@c0d3z3r0 So when using send -e and it fails you want a little nag something like:
"using send **-e** with zstd send might lead to a that is incompatible with the reciever, please try send without the -e argument"

Or a more general warning on incompatible stream:
"Reciever does not seem to be compatible with the send stream, please try and remove the -e argument and try again"

Reply:
ok, will have a look at this in the next days